aten/src/ATen/native/cpu/HistogramKernel.cpp

```diff
@@ -0,0 +1,209 @@
+#include <ATen/native/Histogram.h>
+
+#include <ATen/ATen.h>
+#include <ATen/Dispatch.h>
+#include <ATen/Parallel.h>
+#include <ATen/native/BucketizationUtils.h>
+
+#include <algorithm>
+#include <mutex>
+#include <tuple>
+
+namespace at { namespace native {
+
+namespace {
+
+const int64_t HISTOGRAM_GRAIN_SIZE = 200;
+
+template<typename input_t, typename output_t, typename ExtractWeight_t, bool LinearBinEdges>
+void histogram_cpu_contiguous(Tensor& hist, const Tensor& bin_edges, const Tensor& input, const Scalar& min, const Scalar& max, ExtractWeight_t extract_weight) {
+    int64_t numel_in = input.numel();
+    if (!numel_in) {
+        return;
+    }
+
+    int64_t numel_be = bin_edges.numel();
+    TORCH_CHECK(numel_be > 0, "expected bin_edges to contain at least 1 element but it contains 0");
+
+    const input_t *data_in = input.data_ptr<input_t>();
+    const input_t *data_be = bin_edges.data_ptr<input_t>();
+
+    output_t *data_out = hist.data_ptr<output_t>();
+
+    input_t min_bound = min.to<input_t>();
+    input_t max_bound = max.to<input_t>();
+
+    if (min_bound == max_bound) {
+        min_bound = input.min().item<input_t>();
+        max_bound = input.max().item<input_t>();
+    }
+    if (min_bound == max_bound) {
+        min_bound -= 1;
+        max_bound += 1;
+    }
+
+    const input_t leftmost_bin_edge = data_be[0];
+    const input_t rightmost_bin_edge = data_be[numel_be - 1];
+
+    // we'll process only those elements which fall within the caller-specified [min, max]
+    // range and also fall within the range of the specified bins
+    min_bound = std::max(min_bound, leftmost_bin_edge);
+    max_bound = std::min(max_bound, rightmost_bin_edge);
+
+    std::mutex data_out_mutex;
+    at::parallel_for(0, numel_in, HISTOGRAM_GRAIN_SIZE, [&](int64_t start, int64_t end) {
+        std::vector<output_t> data_out_local(numel_be - 1, output_t(0));
+
+        for (int64_t i = start; i < end; ++i) {
+            if (min_bound <= data_in[i] && data_in[i] <= max_bound) {
+                int64_t pos = -1;
+                if (LinearBinEdges) {
+                    pos = (int64_t)((data_in[i] - leftmost_bin_edge)
+                            / (rightmost_bin_edge - leftmost_bin_edge)
+                            * (numel_be - 1));
+                } else {
+                    pos = std::upper_bound(data_be, data_be + numel_be, data_in[i]) - data_be - 1;
+                }
+
+                // Unlike other bins, the rightmost bin includes its right boundary
+                if (pos == (numel_be - 1))
+                    --pos;
```

Comment:
Style nit: might as well be explicit with `pos -= 1;` since the returned value of the decrement operation is ignored